src/rpc/stats.cpp

```diff
@@ -0,0 +1,203 @@
+    payloadObj.pushKV("avg", payload.avg);
+    payloadObj.pushKV("max", payload.max);
+
+    for (auto const &entry : history) {
+        UniValue historyObj(UniValue::VOBJ);
+        historyObj.pushKV("timestamp", entry.timestamp);
+        historyObj.pushKV("latency", entry.latency);
+        historyObj.pushKV("payload", entry.payload);
+        historyArr.push_back(historyObj);
+    }
+
+    stats.pushKV("name", name);
+    stats.pushKV("count", count);
+    stats.pushKV("lastUsedTime", lastUsedTime);
+    stats.pushKV("latency", latencyObj);
+    stats.pushKV("payload", payloadObj);
+    stats.pushKV("history", historyArr);
+    return stats;
+}
+
+RPCStats RPCStats::fromJSON(UniValue json) {
+    RPCStats stats;
+
+    stats.name = json["name"].get_str();
+    stats.lastUsedTime = json["lastUsedTime"].get_int64();
+    stats.count = json["count"].get_int64();
+    if (!json["latency"].isNull()) {
+        auto latencyObj  = json["latency"].get_obj();
+        stats.latency = {
+            latencyObj["min"].get_int64(),
+            latencyObj["avg"].get_int64(),
+            latencyObj["max"].get_int64()
+        };
+    }
+
+    if (!json["payload"].isNull()) {
+        auto payloadObj  = json["payload"].get_obj();
+        stats.payload = {
+            payloadObj["min"].get_int64(),
+            payloadObj["avg"].get_int64(),
+            payloadObj["max"].get_int64()
+        };
+    }
+
+    if (!json["history"].isNull()) {
+        auto historyArr = json["history"].get_array();
+        for (const auto &entry : historyArr.getValues()) {
+            auto historyObj = entry.get_obj();
+            StatHistoryEntry historyEntry {
+                historyObj["timestamp"].get_int64(),
+                historyObj["latency"].get_int64(),
+                historyObj["payload"].get_int64()
+            };
+            stats.history.push_back(historyEntry);
+        }
+    }
+    return stats;
+}
+
+void CRPCStats::add(const std::string& name, const int64_t latency, const int64_t payload)
+{
+    auto stats = CRPCStats::get(name);
+    if (stats) {
+        stats->count++;
+        stats->lastUsedTime = GetSystemTimeInSeconds();
+        stats->latency = {
+            std::min(latency, stats->latency.min),
+            stats->latency.avg + (latency - stats->latency.avg) / stats->count,
+            std::max(latency, stats->latency.max)
+        };
+        stats->payload = {
+            std::min(payload, stats->payload.min),
+            stats->payload.avg + (payload - stats->payload.avg) / stats->count,
+            std::max(payload, stats->payload.max)
+        };
+    } else {
+        stats = { name, latency, payload };
+    }
+    stats->history.push_back({ stats->lastUsedTime, latency, payload });
+
+    CLockFreeGuard lock(lock_stats);
+    map[name] = *stats;
+}
+
+UniValue CRPCStats::toJSON() {
+    CLockFreeGuard lock(lock_stats);
+
+    UniValue ret(UniValue::VARR);
+    for (auto &[_, stats] : map) {
```

Comment:
Might be better to just copy the map, and do the toJSON conversion outside of the lock.